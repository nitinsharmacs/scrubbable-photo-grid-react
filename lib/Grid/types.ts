import type { GridConfigType, SectionType } from 'lib/types';

export interface GridProps {
  gridData: SectionType[];
  config: GridConfigType;
}
