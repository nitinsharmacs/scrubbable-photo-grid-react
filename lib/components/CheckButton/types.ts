import type { CSSProperties } from 'react';

export interface CheckButtonProps {
  active?: boolean;
  label: string;
  id: string;
  styles?: CSSProperties;
  onClick?: (e: React.SyntheticEvent) => void;
}
