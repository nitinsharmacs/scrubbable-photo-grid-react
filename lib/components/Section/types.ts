import type { SegmentCtx } from 'lib/components/Segment/types';
import type { GridConfigType, SectionConfigType, SectionType } from 'lib/types';

export type SectionSelectionHandler = (
  sectionId: string,
  ctx: SectionCtx
) => void;

export interface SectionProps {
  section: SectionType;
  config: GridConfigType;
  map: SectionConfigType;
  onSelect?: SectionSelectionHandler;
}

export type SectionCtx = {
  selected: boolean;
  segments: {
    [key: string]: SegmentCtx;
  };
};
