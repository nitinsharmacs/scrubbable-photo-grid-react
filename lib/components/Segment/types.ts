import type { SegmentConfig, SegmentType } from 'lib/types';

export interface SegmentProps {
  segment: SegmentType;
  map: SegmentConfig;
}
