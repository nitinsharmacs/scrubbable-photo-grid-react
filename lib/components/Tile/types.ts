import type { ImageType, TileType } from 'lib/types';

export interface TileProps {
  tile: TileType;
  image: ImageType;
}
