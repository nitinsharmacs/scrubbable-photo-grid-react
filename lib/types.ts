export type SectionType = {
  sectionId: string;
  totalImages: number;
  segments: SegmentType[];
  header: string;
};

export type SegmentType = {
  segmentId: string;
  header: string;
  images: ImageType[];
};

export type TileType = {
  width: number;
  height: number;
  top: number;
  left: number;
  aspectRatio?: number;
};

export type ImageType = {
  metadata: {
    width: number;
    height: number;
    orientation?: number;
  };
  imageurl: string;
  imageId: string;
};

export type SegmentConfigType = {
  top: number;
  width: number;
  height: number;
  tiles: TileType[];
  headerHeight: number;
};

export type SegmentsMapType = {
  [key: string]: SegmentConfigType;
};

export type SectionConfigType = {
  height: number;
  width?: number;
  top: number;
  visible: boolean;
  index: number;
  segmentsMap: SegmentsMapType;
  headerHeight: number;
};

export type SectionsMapType = {
  [key: string]: SectionConfigType;
};

export type GridConfigType = {
  containerWidth: number;
  sectionMargin: number;
  segmentMargin: number;
  targetRowHeight: number;
};
