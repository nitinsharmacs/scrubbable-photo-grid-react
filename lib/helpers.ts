import createJustifiedLayout from 'justified-layout';

import type {
  GridConfigType,
  ImageType,
  SectionsMapType,
  SegmentType,
  SegmentsMapType,
  TileType,
} from 'lib/types';

const IMAGE_ASPECT_RATIO: number = 3 / 2;
const COALESCING_FACTOR: number = 7 / 10;

export const estimateSectionHeight = (
  totalImages: number,
  config: GridConfigType
): number => {
  const unwrappedWidth =
    IMAGE_ASPECT_RATIO *
    totalImages *
    config.targetRowHeight *
    COALESCING_FACTOR;
  const rows = Math.ceil(unwrappedWidth / config.containerWidth);
  return rows * config.targetRowHeight;
};

export const createSegmentsMap = (
  segments: SegmentType[],
  gridConfig: GridConfigType
): SegmentsMapType => {
  const map = segments.reduce<SegmentsMapType>(
    (map: SegmentsMapType, segment: SegmentType) => {
      const segmentLayout = createJustifiedLayout(
        segment.images.map((image: ImageType) => image.metadata),
        gridConfig
      );

      const segmentHeight: number = Math.ceil(segmentLayout.containerHeight);

      return {
        ...map,
        [segment.segmentId]: {
          top: gridConfig.segmentMargin + map['prev'].height,
          height: segmentHeight,
          width: gridConfig.containerWidth,
          tiles: segmentLayout.boxes as TileType[],
        },
        prev: {
          ...map.prev,
          height: segmentHeight + gridConfig.segmentMargin + map['prev'].height,
        },
      } as SegmentsMapType;
    },
    { prev: { height: 0, top: 0, width: 0, tiles: [] } }
  );

  delete map['prev'];

  return map;
};

export const updateSectionsTop = (
  sectionsMap: SectionsMapType,
  toBeUpdatedSectionsId: string[],
  delta: number
): SectionsMapType => {
  return toBeUpdatedSectionsId.reduce<SectionsMapType>(
    (map, sectionId: string) => {
      return {
        ...map,
        [sectionId]: { ...map[sectionId], top: map[sectionId].top + delta },
      };
    },
    {
      ...sectionsMap,
    }
  );
};
