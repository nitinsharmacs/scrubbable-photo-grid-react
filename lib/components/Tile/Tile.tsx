import type { TileProps } from 'lib/components/Tile/types';
import { getScaleFactor } from 'lib/helpers';
import './tile.css';
import CheckButton from 'lib/components/CheckButton/CheckButton';

const Tile = ({ tile, image, ctx }: TileProps) => {
  return (
    <div
      className='tile'
      id={image.imageId}
      style={{
        width: `${tile.width}px`,
        height: `${tile.height}px`,
        left: `${tile.left}px`,
        top: `${tile.top}px`,
      }}
    >
      <CheckButton
        active={ctx.selected}
        id={image.imageId}
        label='tile-checkicon'
        styles={{ position: 'absolute', left: '5px', top: '5px' }}
      />

      <figure
        className='tile-image'
        style={
          ctx.selected
            ? {
                transform: `scale(${getScaleFactor(
                  tile.width,
                  30,
                  tile.height,
                  30
                )})`,
              }
            : {}
        }
      >
        <img src={image.imageurl} />
      </figure>
    </div>
  );
};

export default Tile;
