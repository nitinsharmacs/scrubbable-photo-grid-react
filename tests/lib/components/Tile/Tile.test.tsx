import { describe, expect, it, vi, afterEach } from 'vitest';
import React from 'react';
import { render, screen } from '@testing-library/react';
import Tile from 'lib/components/Tile/Tile';
import { ImageType, TileType } from 'lib/types';
import { getScaleFactor } from 'lib/helpers';

vi.mock('lib/helpers', async () => {
  return {
    getScaleFactor: vi.fn().mockReturnValue('scale factor'),
  };
});

afterEach(() => {
  vi.clearAllMocks();
});

describe('Tile', () => {
  it('should match snapshot', () => {
    const tile: TileType = {
      height: 0,
      width: 0,
      top: 0,
      left: 0,
    };
    const image: ImageType = {
      metadata: {
        width: 0,
        height: 0,
        orientation: 0,
      },
      imageurl: 'somehing',
      imageId: 'something1',
    };

    const { container } = render(
      <Tile tile={tile} image={image} ctx={{ selected: false }} />
    );
    expect(container).toMatchSnapshot();
  });

  it('should check tile if selected', () => {
    const tile: TileType = {
      height: 0,
      width: 0,
      top: 0,
      left: 0,
    };
    const image: ImageType = {
      metadata: {
        width: 0,
        height: 0,
        orientation: 0,
      },
      imageurl: 'somehing',
      imageId: 'something1',
    };

    render(<Tile tile={tile} image={image} ctx={{ selected: true }} />);
    expect(screen.getByLabelText('tile-checkicon')).toHaveAttribute(
      'class',
      'checkbutton active'
    );
    expect(screen.getByRole('figure')).toHaveStyle(
      'transform: scale(scale factor)'
    );
    expect(getScaleFactor).toBeCalledWith(0, 30, 0, 30);
  });
});
