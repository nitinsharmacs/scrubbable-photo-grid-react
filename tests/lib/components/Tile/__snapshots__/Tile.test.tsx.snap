// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tile > should match snapshot 1`] = `
<div>
  <div
    class="tile"
    id="something1"
    style="width: 0px; height: 0px; left: 0px; top: 0px;"
  >
    <div
      aria-label="tile-checkicon"
      class="checkbutton "
      data-id="something1"
      style="position: absolute; left: 5px; top: 5px;"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="CheckCircleIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"
        />
      </svg>
    </div>
    <figure
      class="tile-image"
    >
      <img
        src="somehing"
      />
    </figure>
  </div>
</div>
`;
